// @npm/registry-utils/shared-constants.js
'use strict'

const OIDC = {
  ACTOR_TYPE_TRUSTED_PUBLISHER: 'trusted_publisher',
  ACTOR_TYPES: ['trusted_publisher'],
  ACTORS: {
    GITHUB: {
      ID: 'github',
      DISPLAY_NAME: 'GitHub Actions',
      EMAIL: 'npm-oidc-no-reply@github.com',
      // GitHub Actions has a static issuer URL
      ISSUER_URL: 'https://token.actions.githubusercontent.com',
      // https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig
      OPENID_CONFIG_URL: 'https://token.actions.githubusercontent.com/.well-known/openid-configuration',
    },
    GITLAB: {
      ID: 'gitlab',
      DISPLAY_NAME: 'GitLab CI/CD',
      EMAIL: 'npm-oidc-no-reply@github.com',
      ISSUER_URL: 'https://gitlab.com',
      OPENID_CONFIG_URL: 'https://gitlab.com/.well-known/openid-configuration',
    },
    CIRCLECI: {
      ID: 'circleci',
      DISPLAY_NAME: 'CircleCI',
      EMAIL: 'npm-oidc-no-reply@github.com',
      // CircleCI issuer is dynamic: https://oidc.circleci.com/org/<org-id>
      // Provide base URL for pattern matching in consuming code
      ISSUER_URL_BASE: 'https://oidc.circleci.com/org/',
      // OpenID config is also dynamic, constructed from issuer
      // OPENID_CONFIG_URL not applicable — must be derived from issuer at runtime
    },
  },
  // audience URL in OIDC IdP token
  IDP_TOKEN_AUDIENCE_URL_PRODUCTION: 'npm:registry.npmjs.org',
  IDP_TOKEN_AUDIENCE_URL_SANDBOX: 'npm:registry-sandbox.npm.red',
  // issuer URL of OIDC auth token is different for production and staging
  AUTH_TOKEN_ISSUER_URL_PRODUCTION: 'https://registry.npmjs.org',
  AUTH_TOKEN_ISSUER_URL_SANDBOX: 'https://registry-sandbox.npm.red',
}

// `npm-notice` header for the 404 NOT FOUND responses from the registry services
// (registry-frontdoor, npm-auth-ws) and cloudflare-registry when a valid auth token is required to access an endpoint.
const NpmNoticeMessageOn404 = 'Access token expired or revoked. Please try logging in again.'

// Error message to distinguish oidc config NOT FOUND cases from other 404 errors
const OidcNotFoundMessage = 'No valid oidc configs found for package'

module.exports = {
  OIDC,
  NpmNoticeMessageOn404,
  OidcNotFoundMessage,
}
